source/core.py

```diff
@@ -419,6 +419,78 @@ def _handleNVDAModuleCleanupBeforeGUIExit():
 	brailleViewer.destroyBrailleViewer()
 
 
+def _pollForForegroundHWND() -> int:
+	import ui
+	import winUser
+
+	foregroundHWND = winUser.getForegroundWindow()
+	_hasWarned = False
+	WARN_AFTER_SECS = 5
+	MAX_WAIT_TIME_SECS = 20
+	POLL_INTERVAL_SECS = 0.1
+	waitForForegroundHWND_startTime = time.time()
+
+	while not foregroundHWND:
+		# The foreground window can be NULL in certain circumstances,
+		# such as when a window is losing activation.
+		# This should not remain the case for an extended period of time.
+
+		_timeElapsed = time.time() - waitForForegroundHWND_startTime
+		if (
+			not _hasWarned
+			and _timeElapsed > WARN_AFTER_SECS
+		):
+			_hasWarned = True
+			# Allow for braille / speech to be understood before exiting.
+			# Unfortunately we cannot block with a dialog as NVDA cannot read dialogs yet.
+			ui.message(_(
+				# Translators: Message when NVDA is having an issue starting up
+				"NVDA is failing to fetch the foreground window. "
+				"If this continues, NVDA will quit starting in %d seconds." % (MAX_WAIT_TIME_SECS - WARN_AFTER_SECS)
+			))
+
+		if _timeElapsed > MAX_WAIT_TIME_SECS:
+			log.critical("NVDA could not fetch the foreground window. Exiting NVDA.")
+			# Raising exception here causes core.main to exit and NVDA to fail to start
+			raise NVDANotInitializedError("Could not fetch foreground window")
+
+		log.debugWarning("Foreground window not found, fetching again")
+		time.sleep(POLL_INTERVAL_SECS)
+		foregroundHWND = winUser.getForegroundWindow()
+
+	return foregroundHWND

```

Comment:
Looks like I didn't hit submit on this comment.

I want to make sure all the paths here are tested. Doing so manually will be difficult. I think this could be easier to read and unit testable if the polling algorithm was decoupled from the condition and outcomes.

Suggestion:

```python

	# The foreground window can be NULL in certain circumstances,
	# such as when a window is losing activation.
	# This should not remain the case for an extended period of time.
	WARN_AFTER_SECS = 5
	MAX_WAIT_TIME_SECS = 20
	POLL_INTERVAL_SECS = 0.1
	
	success, foregroundHWND = pollForCondition(
		condition=lambda: bool(winUser.getForegroundWindow()),
		maxWait=WARN_AFTER_SECS,
		intervalSeconds=POLL_INTERVAL_SECS
	)
	if not success:
		ui.message(_(
			# Translators: Message when NVDA is having an issue starting up
			"NVDA is failing to fetch the foreground window. "
			"If this continues, NVDA will quit starting in %d seconds." % (MAX_WAIT_TIME_SECS - WARN_AFTER_SECS)
		))
	success, foregroundHWND = pollForCondition(
		condition=lambda: bool(winUser.getForegroundWindow()),
		maxWait=MAX_WAIT_TIME_SECS - WARN_AFTER_SECS,
		intervalSeconds=POLL_INTERVAL_SECS
	)
	if not success:
		log.critical("NVDA could not fetch the foreground window. Exiting NVDA.")
		# Raising exception here causes core.main to exit and NVDA to fail to start
		raise NVDANotInitializedError("Could not fetch foreground window")
	return foregroundHWND
```

Reply:
I have implemented something similar